components/api/core/utils.bal

```diff
@@ -39,3 +40,120 @@ function updatePayloadWithUserInfo (json payload, string field) returns error? {
         log:printDebug(io:sprintf("Response modification failed : User information not found for user : \'%s\'", userId));
     }
 }
+
+function getUserToken(string authorizationHeader, string cookieHeader) returns string | error? {
+    string[] splittedToken = authorizationHeader.split(" ");
+    if splittedToken.length() != 2 || !(splittedToken[0].equalsIgnoreCase(constants:BEARER_HEADER)) {
+        error er = error("Did not receive the token in proper format");
+        return er;
+    }
+    string lastTokenElement = splittedToken[1];
+    string | error firstTokenElement = filter:getCookie(cookieHeader);
+    if (firstTokenElement is error) {
+        error er = error(io:sprintf("Cookie value could not be resolved. %s", firstTokenElement.reason()));
+        return er;
+    } else {
+        string token = io:sprintf("%s%s", firstTokenElement, lastTokenElement);
+        if "" == token {
+            error er = error("Did not receive any token");
+            return er;
+        }
+        return token;
+    }
+}
+
+function deleteArtifactFromRegistry(http:Request deleteArtifactReq, string orgName, string imageName, string artifactVersion) returns error? {
+    string userId = deleteArtifactReq.getHeader(constants:AUTHENTICATED_USER);
+    var token = getUserToken(deleteArtifactReq.getHeader(constants:AUTHORIZATION_HEADER),
+    deleteArtifactReq.getHeader(constants:COOKIE_HEADER));
+
+    if (token is string) {
+        log:printInfo(io:sprintf("Attempting to delete the artifact \'%s/%s:%s\' from the registry", orgName, imageName, artifactVersion));
+        string | error registryScopeForGetManifest = docker_registry:getScopeForFetchManifest(orgName, imageName, artifactVersion, userId, token);
+
+        if (registryScopeForGetManifest is string) {
+            log:printDebug(io:sprintf("Registry scope for getting manifest digest from registry api : %s", registryScopeForGetManifest));
+
+            string | error? tokenToGetManifestDigest = docker_registry:getTokenFromDockerAuth(userId, token, registryScopeForGetManifest);
+
+            if (tokenToGetManifestDigest is string) {
+                log:printDebug("Retrived a token from docker auth to invoke getManifestDigest end point");
+
+                string | error manifestDigest = docker_registry:getManifestDigest(orgName, imageName, artifactVersion, tokenToGetManifestDigest);
+
+                if (manifestDigest is string) {
+                    string | error registryScopeForDeleteManifest = docker_registry:getScopeForDeleteManifest(orgName, imageName, manifestDigest, userId, token);
+
+                    if (registryScopeForDeleteManifest is string) {
+                        log:printDebug(io:sprintf("Registry scope for deleting manifest from registry api: %s", registryScopeForDeleteManifest));
+
+                        string | error? tokenToDeleteManifest = docker_registry:getTokenFromDockerAuth(userId, token, registryScopeForDeleteManifest);
```

Comment:
Can we check whether we can get a token a token with both pull and delete scopes at once ? So that we don't need to call docker auth twice to get two tokens ?

Reply:
It is possible. Just followed the docker registry V2 spec since we thought of implementing a complete ballerina docker client